src/i2c.rs

```diff
@@ -773,6 +810,353 @@ impl<'p> I2CController<'p> {
+        self.regs.smbn_addr5().write(|w| unsafe { w.bits(0) });
+        self.regs.smbn_addr6().write(|w| unsafe { w.bits(0) });
+        self.regs.smbn_addr7().write(|w| unsafe { w.bits(0) });
+        self.regs.smbn_addr8().write(|w| unsafe { w.bits(0) });
+        self.bank_sel(true);
+    }
+
+    fn decode_addr(&mut self) -> Option<u8> {
+        let r = self.regs.smbn_cst().read();
+        if r.gcmatch().bit_is_set() {
+            Some(GENERALCALL_ADDRESS)
+        } else if r.arpmatch().bit_is_set() {
+            Some(ARP_ADDRESS)
+        } else {
+            let r = self.regs.smbn_cst2().read();
+            if r.matcha1f().bit_is_set() {
+                Some(self.regs.smbn_addr1().read().addr().bits())
+            } else if r.matcha2f().bit_is_set() {
+                Some(self.regs.smbn_addr2().read().addr().bits())
+            } else if r.matcha3f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr3().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else if r.matcha4f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr4().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else if r.matcha5f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr5().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else if r.matcha6f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr6().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else if r.matcha7f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr7().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else if self.regs.smbn_cst3().read().matcha8f().bit_is_set() {
+                self.bank_sel(false);
+                let a = self.regs.smbn_addr7().read();
+                self.bank_sel(true);
+                Some(a.addr().bits())
+            } else {
+                None
+            }
+        }
+    }
+
+    async fn handle_listen_transaction(&mut self, mut handler: impl FnMut(u8, ListenCommand)) {
+        let Some(addr) = self.decode_addr() else {
+            // Spurious nmatch, clear it and return
+            self.regs.smbn_st().write(|w| w.nmatch().set_bit());
+            return;
+        };
+
+        let mut buffer = [0u8; GROUP_SIZE as _];
+
+        if self.regs.smbn_st().read().xmit().bit_is_set() {
+            // Read from master
+            self.regs.smbn_st().write(|w| w.nmatch().set_bit());
+
+            let mut bytes_read = 0usize;
+
+            let st = loop {
+                handler(addr, ListenCommand::PrepareRead(&mut buffer));
```

Comment:
What should happen if the user doesn't have enough bytes to fill the buffer? Currently that's not an option I think